wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/ContainerSearch.java

```diff
@@ -0,0 +1,154 @@
+      @Nullable String containerName) {
+    this.parentName = parentName;
+    this.parentSelector = parentSelector;
+    this.containerName = containerName;
+  }
+
+  /**
+   * Searches for containers in the provided list of Kubernetes objects. If any given item in the
+   * list can contain a container (i.e. it is a pod, deployment, etc.) the item is searched for the
+   * containers recursively.
+   *
+   * @param list the list of Kubernetes resources to sift through
+   * @return a list of containers found in the provided object list
+   */
+  public List<Container> search(Collection<? extends HasMetadata> list) {
+    return list.stream()
+        .filter(this::matchMeta)
+        .flatMap(this::findContainers)
+        .filter(this::matchContainer)
+        .collect(toList());
+  }
+
+  private Stream<Container> findContainers(HasMetadata o) {
+    // hopefully, this covers all types of objects that can contain a container
+    if (o instanceof Pod) {
+      return ((Pod) o).getSpec().getContainers().stream();
+    } else if (o instanceof PodTemplate) {
+      return ((PodTemplate) o).getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof DaemonSet) {
+      return ((DaemonSet) o).getSpec().getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof Deployment) {
+      return ((Deployment) o).getSpec().getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof Job) {
+      return ((Job) o).getSpec().getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof ReplicaSet) {
+      return ((ReplicaSet) o).getSpec().getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof ReplicationController) {
+      return ((ReplicationController) o).getSpec().getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof StatefulSet) {
+      return ((StatefulSet) o).getSpec().getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof CronJob) {
+      return ((CronJob) o)
+          .getSpec()
+          .getJobTemplate()
+          .getSpec()
+          .getTemplate()
+          .getSpec()
+          .getContainers()
+          .stream();
+    } else if (o instanceof DeploymentConfig) {
+      return ((DeploymentConfig) o).getSpec().getTemplate().getSpec().getContainers().stream();
+    } else if (o instanceof Template) {
+      return ((Template) o).getObjects().stream().flatMap(this::findContainers);
+    } else {
+      return Stream.empty();
+    }
+  }
+
+  private boolean matchContainer(Container container) {
+    return this.containerName == null || this.containerName.equals(container.getName());
+  }
+
+  private boolean matchMeta(HasMetadata object) {
+    return matches(object.getMetadata(), parentName, parentSelector);
+  }
+
+  private static boolean matches(
+      ObjectMeta metaData, @Nullable String name, @Nullable Map<String, String> labels) {
+    if (name == null) {
```

Comment:
Can not we rewrite it in the following way?
```
return (name == null || matchesByName(metaData, name) 
     && (labels == null || matchesBySelector(metaData, labels);
```
I'm not sure that it will work, but I hope it will. So, try to apply it and unit tests will or won't approve it )